test/api/test-draw.c

```diff
@@ -819,7 +832,7 @@ test_hb_draw_stroking (void)
   /* https://skia-review.googlesource.com/c/skia/+/266945
      https://savannah.nongnu.org/bugs/index.php?57701 */
```

Comment:
The reason `test_hb_draw_stroking` which is now disabled existed was the above discussions.

Reply:
> The reason `test_hb_draw_stroking` which is now disabled existed was the above discussions.
> 
> Update: So this is done intentionally [0f131f5](https://github.com/harfbuzz/harfbuzz/commit/0f131f594bc4aa0e2a81c897b192cd90e3bb759e) just that the reason for this was helping future Skia integration https://skia-review.googlesource.com/c/skia/+/268166/11/src/ports/SkFontHost_FreeType_common.cpp which apparently doesn't matter, maybe you can have Ben's input also.

Can you elaborate?

If a rasterizer cannot accept no-op operations, it's a broken rasterizer. I highly doubt that's what's at fault here, other than cairo asserting if the glyph is fully empty...

@bungeman does Skia have any problem with line-to, conic-to, cubic-to that have all points equal?

Reply:
> If a rasterizer cannot accept no-op operations, it's a broken rasterizer.

Stroking.. I see now how that can be a problem. Still, I think that's best left to such an implementation.

Reply:
Yeah, the Skia thing was mostly about how to handle fonts which use degenerate contours for attachment points (Skia itself handles degenerate segments more or less how SVG specifies, since that's what most of Skia's users expect). These take up no area so are usually harmless, but if you want to stroke the outline (like SVG wants to do) then there are decisions to be made. SVG says degenerate contours get stroked (so draw endcaps) but the reality is that degenerate contours in simple paths in OpenType fonts are used for other reasons. Also, it turns out that FreeType, CoreText, DirectWrite, and even GDI handle the reporting of degenerate contours slightly differently (depends on number of points and if they're on or off curve). So for consistency, Skia's font handling just filters them all out.

That being said, I'm aware of two users of paths. One wants the raw paths with all the original point numbers and whatnot. The other is the more common user who wants to manipulate the path or draw it in some novel way. For these more common users (and this includes Skia) the reporting of any degenerate contours (which can only be meaningful internally to the font) is just noise and needs to be filtered out.

Reply:
I think we want to give as unfiltered data as sensible here, because if we do the filtering at the lower level, higher level code that wants the unfiltered data is left helpless, but the other-way around the higher level code will be able to do the filtering as it fits its needs.

Reply:
> I think we want to give as unfiltered data as sensible here, because if we do the filtering at the lower level, higher level code that wants the unfiltered data is left helpless, but the other-way around the higher level code will be able to do the filtering as it fits its needs.

We're still filtering out the love `move-to`, and repeat move-to's in CFF, so unfortunately that argument doesn't stand. This however, at least takes care of not stroking a single point used as attachment point...

Reply:
> We can add a mode to hb_draw_state_t to choose whether to clean up or not...

Would just add complexity...

Reply:
> We can add a mode to `hb_draw_state_t` to choose whether to clean up or not...

That would be ideal, IMHO.

Reply:
The mode as I see it, would be `allow_empty_path`...

Reply:
In response https://github.com/harfbuzz/harfbuzz/commit/0f131f594bc4aa0e2a81c897b192cd90e3bb759e

> Plus, this throws off point-numbers in uses that rely on it.

I remember CFF/CFF2 didn't have guaranteed point numbers also if I'm not mistaken about what you want to refer, am not insisting on either case anyway.

Reply:
> I think we want to give as unfiltered data as sensible here, because if we do the filtering at the lower level, higher level code that wants the unfiltered data is left helpless, but the other-way around the higher level code will be able to do the filtering as it fits its needs.

So the points we return are not useful for applying native hints... so it's just for things like master-level compatibility-checking, etc.

We are just eliding empty contours, and I'm leaning towards not doing that and leaving it to the consumer. I believe FreeType / cairo do the same. I don't think it belongs to our layer. Otherwise we need to add a new flags parameter to `hb_font_get_glyph_shape` and add that to the `hb_draw_state_t` and change the state implementation.

Reply:
That works for me as well.

Reply:
Humm. There's still this bit of code:
```
  close_path (void *draw_data, hb_draw_state_t &st) 
  { 
    if (st.path_open) 
    { 
      if ((st.path_start_x != st.current_x) || (st.path_start_y != st.current_y)) 
↦       emit_line_to (draw_data, st, st.path_start_x, st.path_start_y); 
      emit_close_path (draw_data, st); 
    } 
```

Can't remove it without leaving spurious `line_to`'s all over the place; if I *don't* remove it, breaks the master-compatibility of outlines. So I feel like a "keep outline compatibility" flag is required in the API somewhere.

Reply:
And really that's the only way I can think of this flag... No further promise I can think of...

Reply:
At any rate, a flags_t in the API seems justified.

Reply:
I'll think some more and make up my mind...

Reply:
So now maybe you also wish to revisit "/* Skip empty contours */" of `hb-ot-glyf-table.hh` and allow emit empty contours emit, it was added here https://github.com/harfbuzz/harfbuzz/commit/3a98c7fae196fd149078af4553b966dc1c8d763b but now the correspond line is https://github.com/harfbuzz/harfbuzz/blob/e59da28/src/hb-ot-glyf-table.hh#L1188

Reply:
Done.  Thanks.